{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "import scipy.io as spio\n",
    "\n",
    "def reverse_complement(seq) :\n",
    "    seq_prime = ''\n",
    "    for j in range(0, len(seq)) :\n",
    "        if seq[j] == 'A' :\n",
    "            seq_prime = 'T' + seq_prime\n",
    "        elif seq[j] == 'C' :\n",
    "            seq_prime = 'G' + seq_prime\n",
    "        elif seq[j] == 'G' :\n",
    "            seq_prime = 'C' + seq_prime\n",
    "        elif seq[j] == 'T' :\n",
    "            seq_prime = 'A' + seq_prime\n",
    "        else :\n",
    "            seq_prime = seq[j] + seq_prime\n",
    "\n",
    "    return seq_prime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        chr      start        end                       gene  reads strand  \\\n",
      "3592  chr10  100007273  100007673  chr10:100007448-100008676      1      -   \n",
      "3593  chr10  100007374  100007774  chr10:100007549-100008676      1      -   \n",
      "4716  chr10  101909674  101910074  chr10:101909849-101911886      1      -   \n",
      "4717  chr10  101911390  101911790  chr10:101911565-101911886      1      -   \n",
      "4946  chr10  101991879  101992279  chr10:101992054-101992982      1      -   \n",
      "\n",
      "               search_region isoform  \n",
      "3592  10:100007273-100007673       A  \n",
      "3593  10:100007374-100007774       B  \n",
      "4716  10:101909674-101910074       A  \n",
      "4717  10:101911390-101911790       B  \n",
      "4946  10:101991879-101992279       A  \n",
      "5312\n"
     ]
    }
   ],
   "source": [
    "\n",
    "emitted_id = []\n",
    "emitted_chrom = []\n",
    "emitted_start = []\n",
    "emitted_end = []\n",
    "emitted_isoform_start = []\n",
    "emitted_isoform_end = []\n",
    "emitted_strand = []\n",
    "emitted_isoform = []\n",
    "emitted_search = []\n",
    "emitted_reads = []\n",
    "\n",
    "i = 0\n",
    "with open('TandemUTR.hg19.gff3') as f:\n",
    "    for line in f:\n",
    "        \n",
    "        if i > 0 :\n",
    "            lineparts = line[:-1].split('\\t')\n",
    "            \n",
    "            chrom = lineparts[0]\n",
    "            event_type = lineparts[2]\n",
    "            start = int(lineparts[3])\n",
    "            end = int(lineparts[4])\n",
    "            strand = lineparts[6]\n",
    "            \n",
    "            id_str = lineparts[8]\n",
    "            \n",
    "            if event_type == 'mRNA' :\n",
    "                emitted_id.append(chrom + ':' + str(start) + '-' + str(end))\n",
    "                \n",
    "                emitted_chrom.append(chrom)\n",
    "                \n",
    "                if strand == '+' :\n",
    "                    emitted_start.append(end - 225)\n",
    "                    emitted_end.append(end + 175)\n",
    "                    \n",
    "                    emitted_search.append(chrom[3:] + ':' + str(end - 225) + '-' + str(end + 175))\n",
    "                else :\n",
    "                    emitted_start.append(start - 175)\n",
    "                    emitted_end.append(start + 225)\n",
    "                    \n",
    "                    emitted_search.append(chrom[3:] + ':' + str(start - 175) + '-' + str(start + 225))\n",
    "                \n",
    "                emitted_isoform_start.append(start)\n",
    "                emitted_isoform_end.append(end)\n",
    "                emitted_strand.append(strand)\n",
    "                \n",
    "                #Prox = B, Dist = A\n",
    "                emitted_isoform.append(id_str.split(';')[0][-1])\n",
    "                \n",
    "                emitted_reads.append(1)\n",
    "        \n",
    "        i += 1\n",
    "\n",
    "\n",
    "bed_df = pd.DataFrame({'chr' : emitted_chrom,\n",
    "                       'start' : emitted_start,\n",
    "                       'end' : emitted_end,\n",
    "                       'gene' : emitted_id,\n",
    "                       'reads' : emitted_reads,\n",
    "                       'strand' : emitted_strand,\n",
    "                       'search_region' : emitted_search,\n",
    "                       'isoform' : emitted_isoform,\n",
    "                })\n",
    "\n",
    "bed_df = bed_df[['chr', 'start', 'end', 'gene', 'reads', 'strand', 'search_region', 'isoform']]\n",
    "bed_df = bed_df.sort_values(by='gene')\n",
    "\n",
    "print(bed_df.head())\n",
    "print(len(bed_df))\n",
    "\n",
    "bed_df.to_csv('Emitted_Tandem_UTR_200up_200dn.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr10  100007273  100007673  chr10:100007448-100008676  1  -  10:100007273-100007673  A\n",
      "chr10  100007374  100007774  chr10:100007549-100008676  1  -  10:100007374-100007774  B\n",
      "chr10  101909674  101910074  chr10:101909849-101911886  1  -  10:101909674-101910074  A\n",
      "chr10  101911390  101911790  chr10:101911565-101911886  1  -  10:101911390-101911790  B\n",
      "chr10  101991879  101992279  chr10:101992054-101992982  1  -  10:101991879-101992279  A\n",
      "\n",
      ">chr10:100007448-100008676\n",
      "GAAATCAGAAGGGGGAACAGTCAGTTTAGTTAAGGATGGAACCTGGGAAAGGCCACCATTCCTGCTTGATGGGGCTCTGATTTGCTCTTGCTCAAGTGGAATAAAACCCCATGGTCTTCTTGACATGATTCTTGATCTTTTCTCCACTGAGACACACTTAAGTGATGATCCTTACAGGACTGACACCCTAATGCCAATAAAAGTTGCTCATTATGGACTGCTACAAAGACCAGACCAGCTGTGATTATGTACACACCCCAGAACCACAGACCTACCCTTTTGCCTCCCATGGAGCCAGGGCCTGGGATACACAGGGCACACATGCTTCTCATGACTGCCCATCTACATTTCTCCAATCAGACATAAGTCTGCTTCCCAGTCAGCCTCCCCAGGCTTAGAG\n",
      ">chr10:100007549-100008676\n",
      "ATTGAATAAATTTTTGCCATGGAAAGAACATCAAACAAGCCACTCATCTCTACAGAGATAAGAAAACAAGTTTGGCAGAGCAAGAGACAGAAGACCGTGGAGAAATCAGAAGGGGGAACAGTCAGTTTAGTTAAGGATGGAACCTGGGAAAGGCCACCATTCCTGCTTGATGGGGCTCTGATTTGCTCTTGCTCAAGTGGAATAAAACCCCATGGTCTTCTTGACATGATTCTTGATCTTTTCTCCACTGAGACACACTTAAGTGATGATCCTTACAGGACTGACACCCTAATGCCAATAAAAGTTGCTCATTATGGACTGCTACAAAGACCAGACCAGCTGTGATTATGTACACACCCCAGAACCACAGACCTACCCTTTTGCCTCCCATGGAGCCAGG\n",
      ">chr10:101909849-101911886\n",
      "CGGGGCTGAGGAACAGGGAAATGCCGCTGTGAAGTCTTAAAGCACTTCTGCTTAAACTCCCATGTGTGAGGAGTGTGCCTCCCTGTGCCCTCTCAGCTCTGAGGCTGGCCGTCTTTCGGGGTGTTCCTTTTGGCAAATATACACTGTAATCTTGAGTCTAAATTTATATGTTGAAATGCTACCttttttaaaataagaaactaaataaaattattttaCTATCAGTGATTTGTGTTTTTTTTTCCCACGTCTTCATTTTTCTGTCATCTTAACATGCTCCTAAGGCTGTCTTGCACATTTCAACTAGGGAGCCAGTTTATTTGCGGGTTTAACTTAACTGTAGAGGCTTTTGTTATACTCTTGCCAGAATACTGCCTACCTGGTATAAGAATGTTTAAAA\n",
      ">chr10:101911565-101911886\n",
      "ACTTCATCTGTTCCACCTCTCCTGCGATAGTCCTGGGTGCTCCACTGATTGGAGGATAGAGCCAGCTGTCTGACACACAAATGGTCTTTTCAGCCACAGTCTTATCAAGTATCCTATATGTATTCCTTTCTAAACTGCTACTCATGAATGAGGAAAGTCTGATGCTAAGATACTGCCTGCACTGGAATGTTAAACACTAAATATATAACAAGCTGTGTTTTCCTAAGCTGAGATCTGTTGAATAATGTTTACATTCGTCCCCCGGGGAAATGTATGCTCAGCCACCATTCAAGAGATGACTGAGAAGGAGATGGTAAGTTCAAGAAGACTGATTGCACCTGGGACCCAGGCCCTTTCTTTGGGATCCAGTCCCAGCCTTCATCCATGTGATTAAGATCCA\n",
      ">chr10:101992054-101992982\n",
      "gagaattgcttgagcctgggaggtggaggttgcagtgagccaagatcgctccattgcactccagcctgggtgacagtgaaactccatctcaaaaataaaagaataaaaGTATGTCTGTCATCCAGCTCCTATGTCTGTTATCCAGCTCCAAGTACAGCTTGTGTATATCAACATTTTCAAAAACCTTTAAACTACCAAATGAAATCCAGATGTTTTTCCATGGTTGAGAAGTGTCTTGTGTTCATGTTGCCTGTGACCGTTGAAATGAAGTAGTATGTGTAATAGCAGGGGCTAGAGCTTTAGGCCCAGCTGTTACTCCAAAATCACACAGGGCAGTTTCTCCCTGAACAACCCAAAGTGGAAAAGGGATCCAAGGCTTTTGCTATTTTTCTCTTAAGCA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "hg19_fai     = '../apadb/hg19.fa.fai'\n",
    "hg19_fa      = '../apadb/hg19.fa'\n",
    "\n",
    "# fasta\n",
    "output_fa = 'Emitted_Tandem_UTR_200up_200dn_Seqs.fa'\n",
    "#!bedtools getfasta -name -s -fi \"$hg19_fa\" -bed \"$output_bed\" -fo | cut -d : -f-4 > \"$output_fa\"\n",
    "!bedtools getfasta -name -s -fi \"$hg19_fa\" -bed \"Emitted_Tandem_UTR_200up_200dn.bed\" -fo \"$output_fa\"\n",
    "    \n",
    "# file tops\n",
    "!head -5 \"Emitted_Tandem_UTR_200up_200dn.bed\" | column -t ; echo\n",
    "!head -10 \"$output_fa\" ; echo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5312\n",
      "{'end': 102589873, 'seq': 'GGCGGCGAGGGCGGCGAGGGCGCCGAGGTCCGGCCCATCCCAGTCCTGTGGGGCTGGCCGGGCAGAGACCCCGGACCCAGGCCCAGGCCTAACCTGCTAAATGTCCCCGGACGGTTCTGGTCTCCTCGGCCACTTTCAGTGCGTCGGTTCGTTTTGATTCTTTTTCTTTTGTGCACATAAGAAATAAATAATAATAATAAATAAAGAATAAAATTTTGTATGTCACTCCCCATGGCTCCAAGTTTGTCTCTCCCTGTCTCTGAGATGGGCCTCCCCTCCATTGGTCGATCCCCAAAAGCCCCTTCAATGATCCTCCCAACTACACTCCCGCTGCCACCTCCAACTCCTTTGCCGAGACCCCCTTGGCGGCAGCTGAACCACGGCGAAGGCCGAGACTAGC', 'chrom': 'chr10', 'isoform': 'A', 'start': 102589473, 'seq_map': {}, 'var': {'samples': {}}, 'ref': {'samples': {}}, 'strand': '+'}\n"
     ]
    }
   ],
   "source": [
    "#Inflate sample whitelist\n",
    "sample_set = {}\n",
    "i = 0\n",
    "with open('E-GEUV-1.sdrf.txt') as f:\n",
    "    for line in f:\n",
    "        \n",
    "        if i > 0 :\n",
    "            lineparts = line[:-1].split('\\t')\n",
    "            sample_set[lineparts[0]] = True\n",
    "\n",
    "        i += 1\n",
    "\n",
    "#Inflate Tandem UTR events\n",
    "event_dict = {}\n",
    "with open('Emitted_Tandem_UTR_200up_200dn.bed') as f:\n",
    "    for line in f:\n",
    "        lineparts = line[:-1].split('\\t')\n",
    "        event_id = lineparts[3]\n",
    "        \n",
    "        event_dict[event_id] = {}\n",
    "        event_dict[event_id]['chrom'] = lineparts[0]\n",
    "        event_dict[event_id]['start'] = int(lineparts[1])\n",
    "        event_dict[event_id]['end'] = int(lineparts[2])\n",
    "        event_dict[event_id]['strand'] = lineparts[5]\n",
    "        event_dict[event_id]['isoform'] = lineparts[7]\n",
    "        \n",
    "        event_dict[event_id]['ref'] = {}\n",
    "        event_dict[event_id]['ref']['samples'] = {}\n",
    "        event_dict[event_id]['var'] = {}\n",
    "        event_dict[event_id]['var']['samples'] = {}\n",
    "        \n",
    "        event_dict[event_id]['seq_map'] = {}\n",
    "\n",
    "i = 0\n",
    "with open('Emitted_Tandem_UTR_200up_200dn_Seqs.fa') as f:\n",
    "    event_id = ''\n",
    "    for line in f:\n",
    "        \n",
    "        linep = line[:-1]\n",
    "        \n",
    "        if i % 2 == 0 :\n",
    "            event_id = linep[1:]\n",
    "        else :\n",
    "            event_dict[event_id]['seq'] = linep.upper()\n",
    "        \n",
    "        i += 1\n",
    "\n",
    "\n",
    "print(len(event_dict))\n",
    "print(event_dict['chr10:102587376-102589698'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def add_snp(seq, strand, var_type, ref, var, start_pos, var_pos) :\n",
    "    \n",
    "    rel_pos_start = var_pos - start_pos - 1\n",
    "    rel_pos_end = rel_pos_start + len(ref)\n",
    "    if rel_pos_start < 5 or rel_pos_end > 395 :\n",
    "        return '', 0\n",
    "    \n",
    "    var_seq = seq[:]\n",
    "    rel_pos = 0\n",
    "    if strand == '-' :\n",
    "        var_seq = reverse_complement(var_seq)\n",
    "    \n",
    "    if var_type == 'SNP' :\n",
    "        rel_pos = var_pos - start_pos - 1\n",
    "        if var_seq[rel_pos] == ref and rel_pos >= 0:\n",
    "            var_seq = var_seq[0:rel_pos] + var + var_seq[rel_pos+1:]\n",
    "        elif rel_pos != -1 :\n",
    "            print(seq)\n",
    "            print(rel_pos)\n",
    "            print(strand)\n",
    "            print(ref)\n",
    "            print(var)\n",
    "            \n",
    "            print('ERROR (SNP): Sequence not aligned with genome reference.')\n",
    "            return '', -1\n",
    "    #else :\n",
    "    #    return '', -1\n",
    "    elif var_type == 'INDEL' :\n",
    "        rel_pos_start = var_pos - start_pos - 1\n",
    "        rel_pos_end = rel_pos_start + len(ref)\n",
    "        \n",
    "        rel_pos = rel_pos_start\n",
    "        \n",
    "        if var_seq[rel_pos_start:rel_pos_end] == ref :\n",
    "            var_seq = var_seq[0:rel_pos_start] + var + var_seq[rel_pos_end:]\n",
    "        else :\n",
    "            \n",
    "            print(var_seq)\n",
    "            print(rel_pos_start)\n",
    "            print(rel_pos_end)\n",
    "            print(var_seq[rel_pos_start:rel_pos_end])\n",
    "            print(ref)\n",
    "            print(var)\n",
    "            \n",
    "            print('ERROR (INDEL): Sequence not aligned with genome reference.')\n",
    "            print('' + 1)\n",
    "    elif var_type == 'OTHER' and var == '<DEL>' :\n",
    "        rel_pos_start = var_pos - start_pos - 1\n",
    "        rel_pos_end = rel_pos_start + len(ref)\n",
    "        \n",
    "        rel_pos = rel_pos_start\n",
    "        \n",
    "        if var_seq[rel_pos_start:rel_pos_end] == ref :\n",
    "            var_seq = var_seq[0:rel_pos_start] + var_seq[rel_pos_end:]\n",
    "        else :\n",
    "            print('ERROR (DEL): Sequence not aligned with genome reference.')\n",
    "            print('' + 1)\n",
    "    #elif var_type == 'OTHER' and ref == '<INS>' :\n",
    "    #    rel_pos = var_pos - start_pos - 1\n",
    "    #    var_seq = var_seq[0:rel_pos] + var + var_seq[rel_pos:]\n",
    "    else :\n",
    "        return '', 0\n",
    "    \n",
    "    \n",
    "    var_seq += ('X' * 20)\n",
    "    var_seq = var_seq[:400]\n",
    "    \n",
    "    if strand == '-' :\n",
    "        var_seq = reverse_complement(var_seq)\n",
    "        rel_pos = 400 - (rel_pos + 1)\n",
    "        \n",
    "    return var_seq, rel_pos\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 1 events.\n",
      "Processed 1001 events.\n",
      "Processed 2001 events.\n",
      "Processed 3001 events.\n",
      "Processed 4001 events.\n",
      "Processed 5001 events.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#Inflate sample variant calls\n",
    "\n",
    "valid_sample_dict = {}\n",
    "\n",
    "event_i = 0\n",
    "for event_id in event_dict :\n",
    "    \n",
    "    seq = event_dict[event_id]['seq']\n",
    "    chrom = event_dict[event_id]['chrom']\n",
    "    start = event_dict[event_id]['start']\n",
    "    end = event_dict[event_id]['end']\n",
    "    strand = event_dict[event_id]['strand']\n",
    "    \n",
    "    valid_sample_dict[event_id] = {}\n",
    "    \n",
    "    call_file = 'snps2/' + event_id.replace(':', '_') + '_' + chrom[3:] + '_' + str(start) + '-' + str(end) + '.txt'\n",
    "    try :\n",
    "        with open(call_file) as f:\n",
    "            for line in f:\n",
    "                lineparts = line[:-1].split('\\t')\n",
    "\n",
    "                snp_type = lineparts[0]\n",
    "                snp_pos = int(lineparts[2])\n",
    "\n",
    "                ref = lineparts[3]\n",
    "                var = lineparts[4]\n",
    "\n",
    "                if len(ref) > 10 or len(var) > 10 :\n",
    "                    continue\n",
    "\n",
    "                for sample_index in range(5, len(lineparts)) :\n",
    "                    sample_lineparts = lineparts[sample_index].split('=')\n",
    "\n",
    "                    sample = sample_lineparts[0]\n",
    "\n",
    "                    alleles = sample_lineparts[1].split('|')\n",
    "                    if len(alleles) == 1 :\n",
    "                        alleles = sample_lineparts[1].split('/')\n",
    "                    if len(alleles) == 1 :\n",
    "                        continue\n",
    "\n",
    "                    allele1 = int(alleles[0])\n",
    "                    allele2 = int(alleles[1])\n",
    "\n",
    "                    if sample not in sample_set :\n",
    "                        continue\n",
    "\n",
    "                    valid_sample_dict[event_id][sample] = True\n",
    "\n",
    "                    zyg = ''\n",
    "                    if allele1 == 0 and allele2 == 0 :\n",
    "                        continue\n",
    "                    elif allele1 > 0 and allele2 > 0 :\n",
    "                        zyg = 2\n",
    "                    else :\n",
    "                        zyg = 1\n",
    "\n",
    "                    if sample not in event_dict[event_id]['var']['samples'] :\n",
    "                        event_dict[event_id]['var']['samples'][sample] = {}\n",
    "                        event_dict[event_id]['var']['samples'][sample]['seq'] = seq\n",
    "                        event_dict[event_id]['var']['samples'][sample]['zyg'] = 2\n",
    "                        event_dict[event_id]['var']['samples'][sample]['count'] = 0\n",
    "                        event_dict[event_id]['var']['samples'][sample]['type'] = ''\n",
    "                        event_dict[event_id]['var']['samples'][sample]['pos'] = ''\n",
    "                        event_dict[event_id]['var']['samples'][sample]['snpid'] = ''\n",
    "\n",
    "                    add_mut = False\n",
    "                    if event_dict[event_id]['var']['samples'][sample]['count'] == 0 :\n",
    "                        add_mut = True\n",
    "                    if snp_type == 'SNP' and 'OTHER' not in event_dict[event_id]['var']['samples'][sample]['type'] and 'INDEL' not in event_dict[event_id]['var']['samples'][sample]['type'] :\n",
    "                        add_mut = True\n",
    "                    \n",
    "                    if add_mut == True :\n",
    "                        var_seq, rel_pos = add_snp(event_dict[event_id]['var']['samples'][sample]['seq'], strand, snp_type, ref, var, start, snp_pos)\n",
    "                        if var_seq != '' :\n",
    "                            event_dict[event_id]['var']['samples'][sample]['seq'] = var_seq\n",
    "                            event_dict[event_id]['var']['samples'][sample]['zyg'] = min(zyg, event_dict[event_id]['var']['samples'][sample]['zyg'])\n",
    "                            event_dict[event_id]['var']['samples'][sample]['count'] += 1\n",
    "                            if event_dict[event_id]['var']['samples'][sample]['type'] == '' :\n",
    "                                event_dict[event_id]['var']['samples'][sample]['type'] = snp_type\n",
    "                            else :\n",
    "                                event_dict[event_id]['var']['samples'][sample]['type'] += ',' + snp_type\n",
    "                            if event_dict[event_id]['var']['samples'][sample]['pos'] == '' :\n",
    "                                event_dict[event_id]['var']['samples'][sample]['pos'] = str(rel_pos)\n",
    "                            else :\n",
    "                                event_dict[event_id]['var']['samples'][sample]['pos'] += ',' + str(rel_pos)\n",
    "                            \n",
    "                            if event_dict[event_id]['var']['samples'][sample]['snpid'] == '' :\n",
    "                                event_dict[event_id]['var']['samples'][sample]['snpid'] = str(chrom) + str(strand) + ':' + str(int(start)) + '-' + str(int(end)) + ':' + str(int(snp_pos)) + '/' + str(snp_type) + '/' + str(ref) + '/' + str(var)\n",
    "                            else :\n",
    "                                event_dict[event_id]['var']['samples'][sample]['snpid'] += ',' + str(chrom) + str(strand) + ':' + str(int(start)) + '-' + str(int(end)) + ':' + str(int(snp_pos)) + '/' + str(snp_type) + '/' + str(ref) + '/' + str(var)\n",
    "\n",
    "        #print('Number of variant samples for event ' + event_id + ': ' + str(len(event_dict[event_id]['var']['samples'])))\n",
    "\n",
    "        if event_i % 1000 == 0 :\n",
    "            print('Processed ' + str(event_i + 1) + ' events.')\n",
    "    except IOError :\n",
    "        print('ERROR: Could not open file: ' + call_file)\n",
    "        \n",
    "    event_i += 1\n",
    "\n",
    "\n",
    "for event_id in event_dict :\n",
    "    for sample in sample_set :\n",
    "        \n",
    "        if sample not in event_dict[event_id]['var']['samples'] and sample in valid_sample_dict[event_id] :\n",
    "            event_dict[event_id]['ref']['samples'][sample] = {}\n",
    "        \n",
    "        elif sample in valid_sample_dict[event_id] :\n",
    "            var_event = event_dict[event_id]['var']['samples'][sample]\n",
    "            \n",
    "            if var_event['seq'] not in event_dict[event_id]['seq_map'] :\n",
    "                event_dict[event_id]['seq_map'][var_event['seq']] = {}\n",
    "            event_dict[event_id]['seq_map'][var_event['seq']][sample] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'end': 34019781, 'seq': 'ATAAACTTTAATAGTCATAGAATTGTAAATCACTATGGTTAACAGAAAGTGAAAATATTTTCATGCAGATGATGTGAACAGCCATGTGAATAGGTGACTTGGGCACACAGCAGGGTCATATGACTTCAGAAAACTTCGCTTTTCAGTTATTCCATTGTTATAATGTCAACCCTTTAAGACATTGATGTTTAGAGGGCTCACAAATAAAATCTGAATACCTGTAAGGAAAGAGGTTTTTTATCACATACCTTAAGTCTTTGTAATGTTCATGCTTAAATTCTAAGTTTTCACCTTAGTGACACACAAGGTTTGGTTGTAGGCAACAAGTCCCAGGTGTGTGGGAAATTGATTCACAACAGAGATGGGAAAAGGTGCAGATAATTTCCAATGCCTTCACAAT', 'chrom': 'chr5', 'isoform': 'A', 'start': 34019381, 'seq_map': {}, 'var': {'samples': {}}, 'ref': {'samples': {}}, 'strand': '-'}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(event_dict['chr5:34019556-34020686'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "#Inflate MISO expression\n",
    "\n",
    "for sample in sample_set :\n",
    "    \n",
    "    i = 0\n",
    "    with open('geuvadis/' + sample + '_summary/summary/geuvadis_output.miso_summary') as f:\n",
    "        for line in f:\n",
    "\n",
    "            if i > 0 :\n",
    "                lineparts = line[:-1].split('\\t')\n",
    "                \n",
    "                psi_mean = float(lineparts[1])\n",
    "                psi_low = float(lineparts[2])\n",
    "                psi_high = float(lineparts[3])\n",
    "                \n",
    "                chrom = lineparts[7]\n",
    "                \n",
    "                start_positions = lineparts[9].split(',')\n",
    "                end_positions = lineparts[10].split(',')\n",
    "                \n",
    "                a_event_id = chrom + \":\" + start_positions[0] + '-' + end_positions[0]\n",
    "                b_event_id = chrom + \":\" + start_positions[1] + '-' + end_positions[1]\n",
    "                \n",
    "                if a_event_id in event_dict and sample in event_dict[a_event_id]['var']['samples'] :\n",
    "                    event_dict[a_event_id]['var']['samples'][sample]['psi_mean'] = psi_mean\n",
    "                    event_dict[a_event_id]['var']['samples'][sample]['psi_low'] = psi_low\n",
    "                    event_dict[a_event_id]['var']['samples'][sample]['psi_high'] = psi_high\n",
    "                elif a_event_id in event_dict and sample in event_dict[a_event_id]['ref']['samples'] :\n",
    "                    event_dict[a_event_id]['ref']['samples'][sample]['psi_mean'] = psi_mean\n",
    "                    event_dict[a_event_id]['ref']['samples'][sample]['psi_low'] = psi_low\n",
    "                    event_dict[a_event_id]['ref']['samples'][sample]['psi_high'] = psi_high\n",
    "                \n",
    "                if b_event_id in event_dict and sample in event_dict[b_event_id]['var']['samples'] :\n",
    "                    event_dict[b_event_id]['var']['samples'][sample]['psi_mean'] = 1.0 - psi_mean\n",
    "                    event_dict[b_event_id]['var']['samples'][sample]['psi_low'] = 1.0 - psi_high\n",
    "                    event_dict[b_event_id]['var']['samples'][sample]['psi_high'] = 1.0 - psi_low\n",
    "                elif b_event_id in event_dict and sample in event_dict[b_event_id]['ref']['samples'] :\n",
    "                    event_dict[b_event_id]['ref']['samples'][sample]['psi_mean'] = 1.0 - psi_mean\n",
    "                    event_dict[b_event_id]['ref']['samples'][sample]['psi_low'] = 1.0 - psi_high\n",
    "                    event_dict[b_event_id]['ref']['samples'][sample]['psi_high'] = 1.0 - psi_low\n",
    "                \n",
    "\n",
    "            i += 1\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "#Filter variant events\n",
    "\n",
    "ci_limit = 0.25#0.25\n",
    "\n",
    "for event_id in event_dict :\n",
    "    \n",
    "    delete_list = []\n",
    "    for sample in event_dict[event_id]['var']['samples'] :\n",
    "        if 'psi_mean' not in event_dict[event_id]['var']['samples'][sample] :\n",
    "            delete_list.append(sample)\n",
    "            continue\n",
    "        if event_dict[event_id]['var']['samples'][sample]['psi_high'] - event_dict[event_id]['var']['samples'][sample]['psi_low'] > ci_limit :\n",
    "            delete_list.append(sample)\n",
    "            continue\n",
    "        \n",
    "        if event_dict[event_id]['var']['samples'][sample]['count'] == 0 :\n",
    "            delete_list.append(sample)\n",
    "            continue\n",
    "        \n",
    "    for sample in delete_list :\n",
    "        del event_dict[event_id]['var']['samples'][sample]\n",
    "    \n",
    "    delete_list = []\n",
    "    for sample in event_dict[event_id]['ref']['samples'] :\n",
    "        if 'psi_mean' not in event_dict[event_id]['ref']['samples'][sample] :\n",
    "            delete_list.append(sample)\n",
    "            continue\n",
    "        if event_dict[event_id]['ref']['samples'][sample]['psi_high'] - event_dict[event_id]['ref']['samples'][sample]['psi_low'] > ci_limit :\n",
    "            delete_list.append(sample)\n",
    "            continue\n",
    "        \n",
    "    for sample in delete_list :\n",
    "        del event_dict[event_id]['ref']['samples'][sample]\n",
    "\n",
    "\n",
    "min_ref_samples = 5\n",
    "min_var_samples = 1\n",
    "\n",
    "delete_list = []\n",
    "for event_id in event_dict :\n",
    "    for seq in event_dict[event_id]['seq_map'] :\n",
    "        delete_list_seq = []\n",
    "        for sample in event_dict[event_id]['seq_map'][seq] :\n",
    "            if sample not in event_dict[event_id]['var']['samples'] :\n",
    "                delete_list_seq.append(sample)\n",
    "        for sample in delete_list_seq :\n",
    "            del event_dict[event_id]['seq_map'][seq][sample]\n",
    "    \n",
    "    if len(event_dict[event_id]['var']['samples']) <= min_var_samples :\n",
    "        delete_list.append(event_id)\n",
    "    elif len(event_dict[event_id]['ref']['samples']) <= min_ref_samples :\n",
    "        delete_list.append(event_id)\n",
    "\n",
    "for event_id in delete_list :\n",
    "    del event_dict[event_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1945\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(len(event_dict))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#Make Valid PAS lookup hierarchy\n",
    "\n",
    "cano_pas1 = 'AATAAA'\n",
    "cano_pas2 = 'ATTAAA'\n",
    "\n",
    "valid_pas = []\n",
    "\n",
    "valid_pas.append({})\n",
    "valid_pas[0]['AATAAA'] = True\n",
    "valid_pas[0]['ATTAAA'] = True\n",
    "\n",
    "valid_pas.append({})\n",
    "valid_pas[1]['AGTAAA'] = True\n",
    "valid_pas[1]['TATAAA'] = True\n",
    "valid_pas[1]['CATAAA'] = True\n",
    "valid_pas[1]['GATAAA'] = True\n",
    "\n",
    "valid_pas.append({})\n",
    "for pos in range(0, 6) :\n",
    "    for base in ['A', 'C', 'G', 'T'] :\n",
    "        valid_pas[2][cano_pas1[:pos] + base + cano_pas1[pos+1:]] = True\n",
    "\n",
    "valid_pas.append({})\n",
    "for pos1 in range(0, 6) :\n",
    "    for pos2 in range(pos1 + 1, 6) :\n",
    "        for base1 in ['A', 'C', 'G', 'T'] :\n",
    "            for base2 in ['A', 'C', 'G', 'T'] :\n",
    "                valid_pas[3][cano_pas1[:pos1] + base1 + cano_pas1[pos1+1:pos2] + base2 + cano_pas1[pos2+1:]] = True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def get_average_ref_psi(event) :\n",
    "    psi_mean = 0.0\n",
    "    psi_mean_count = 0.0\n",
    "    \n",
    "    any_member = None\n",
    "    \n",
    "    for sample in event['ref']['samples'] :\n",
    "        if event['ref']['samples'][sample]['psi_mean'] >= 0 :\n",
    "            psi_mean += event['ref']['samples'][sample]['psi_mean']\n",
    "            psi_mean_count += 1.0\n",
    "            any_member = event['ref']['samples'][sample]\n",
    "    \n",
    "    return psi_mean / psi_mean_count, psi_mean_count, any_member\n",
    "\n",
    "def get_average_var_psi(event_id, event, seq, zyg) :\n",
    "    psi_mean = 0.0\n",
    "    psi_mean_count = 0.0\n",
    "    \n",
    "    any_member = None\n",
    "    \n",
    "    for sample in event['seq_map'][seq] :\n",
    "        if event['var']['samples'][sample]['psi_mean'] >= 0 and event['var']['samples'][sample]['zyg'] == zyg :\n",
    "            psi_mean += event['var']['samples'][sample]['psi_mean']\n",
    "            psi_mean_count += 1.0\n",
    "            any_member = event['var']['samples'][sample]\n",
    "    \n",
    "    if psi_mean_count <= 0 :\n",
    "        return -1, 0, None\n",
    "    \n",
    "    return psi_mean / psi_mean_count, psi_mean_count, any_member\n",
    "\n",
    "def align_seqs(ref_seq, var_seq, cut_start, cut_end, before_cut = 35, after_cut = 5) :\n",
    "    \n",
    "    align_j = cut_start - 25\n",
    "    aligned = -1\n",
    "    \n",
    "    for i in range(0, len(valid_pas)) :\n",
    "        for j in range(cut_start - before_cut, cut_start + after_cut) :\n",
    "\n",
    "            candidate_pas_ref = ref_seq[j:j+6]\n",
    "            candidate_pas_var = var_seq[j:j+6]\n",
    "\n",
    "            if candidate_pas_ref in valid_pas[i] or candidate_pas_var in valid_pas[i] :\n",
    "                align_j = j\n",
    "                aligned = i\n",
    "                break\n",
    "        if aligned != -1 :\n",
    "            break\n",
    "    \n",
    "    aligned_ref_seq = (ref_seq[align_j-50:])[:186]\n",
    "    aligned_var_seq = (var_seq[align_j-50:])[:186]\n",
    "    \n",
    "    return aligned_ref_seq, aligned_var_seq, aligned, get_mut_pos(aligned_ref_seq, aligned_var_seq)\n",
    "\n",
    "def get_mut_pos(ref_seq, var_seq) :\n",
    "    mut_pos = ''\n",
    "    for j in range(0, len(ref_seq)) :\n",
    "        if ref_seq[j] != var_seq[j] :\n",
    "            mut_pos += str(j) + ','\n",
    "    return mut_pos[:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "#Deflate data set\n",
    "\n",
    "with open('APA_Tandem_UTR_GEUV_With_Id.csv', 'w') as out_f :\n",
    "    \n",
    "    out_f.write('snp_id' + '\\t' + 'snp_type' + '\\t' + 'isoform' + '\\t' + 'zyg' + '\\t' + 'pas' + '\\t' + 'ref_seq' + '\\t' + 'var_seq' + '\\t' + 'ref_ratio' + '\\t' + 'var_ratio' + '\\t' 'diff' + '\\t' 'diff_logodds' + '\\t' + 'snp_count' + '\\t' + 'snp_pos' + '\\t' + 'ref_samples' + '\\t' + 'var_samples' + '\\n')\n",
    "    \n",
    "    for event_id in event_dict :\n",
    "        \n",
    "        ref_seq = event_dict[event_id]['seq']\n",
    "        ref_psi, ref_count, ref_member = get_average_ref_psi(event_dict[event_id])\n",
    "        \n",
    "        isoform = event_dict[event_id]['isoform']\n",
    "        \n",
    "        for var_seq in event_dict[event_id]['seq_map'] :\n",
    "            \n",
    "            aligned_ref_seq, aligned_var_seq, aligned, mut_pos = align_seqs(ref_seq, var_seq, 225, 225+1)\n",
    "            \n",
    "            #HETEROZYGOUS VARIANT\n",
    "            var_psi, var_count, var_member = get_average_var_psi(event_id, event_dict[event_id], var_seq, 1)\n",
    "            \n",
    "            if var_psi >= 0.0 and var_psi <= 1.0 and aligned != -1 and mut_pos != '' :\n",
    "                psi_limit = 0.15\n",
    "                var_sample_limit = 5\n",
    "                ref_sample_limit = 10\n",
    "\n",
    "                if np.abs(var_psi - ref_psi) >= psi_limit and var_count >= var_sample_limit and ref_count >= ref_sample_limit :\n",
    "                    diff_logodds = str(round(np.log(var_psi / (1.0 - var_psi)) - np.log(ref_psi / (1.0 - ref_psi)), 2))\n",
    "                    out_f.write(var_member['snpid'] + '\\t' + var_member['type'] + '\\t' + isoform + '\\t' + 'HET' + '\\t' + str(aligned) + '\\t' + aligned_ref_seq + '\\t' + aligned_var_seq + '\\t' + str(ref_psi) + '\\t' + str(var_psi) + '\\t' + str(var_psi - ref_psi) + '\\t' + diff_logodds + '\\t' + str(var_member['count']) + '\\t' + str(mut_pos) + '\\t' + str(ref_count) + '\\t' + str(var_count) + '\\n')\n",
    "                elif np.abs(var_psi - ref_psi) >= 0.3 and var_count >= 3 and ref_count >= ref_sample_limit :\n",
    "                    diff_logodds = str(round(np.log(var_psi / (1.0 - var_psi)) - np.log(ref_psi / (1.0 - ref_psi)), 2))\n",
    "                    out_f.write(var_member['snpid'] + '\\t' + var_member['type'] + '\\t' + isoform + '\\t' + 'HET' + '\\t' + str(aligned) + '\\t' + aligned_ref_seq + '\\t' + aligned_var_seq + '\\t' + str(ref_psi) + '\\t' + str(var_psi) + '\\t' + str(var_psi - ref_psi) + '\\t' + diff_logodds + '\\t' + str(var_member['count']) + '\\t' + str(mut_pos) + '\\t' + str(ref_count) + '\\t' + str(var_count) + '\\n')\n",
    "        \n",
    "            #HOMOZYGOUS VARIANT\n",
    "            var_psi, var_count, var_member = get_average_var_psi(event_id, event_dict[event_id], var_seq, 2)\n",
    "            \n",
    "            if var_psi >= 0.0 and var_psi <= 1.0 and aligned != -1 and mut_pos != '' :\n",
    "                psi_limit = 0.10\n",
    "                var_sample_limit = 5\n",
    "                ref_sample_limit = 10\n",
    "                \n",
    "                if np.abs(var_psi - ref_psi) >= psi_limit and var_count >= var_sample_limit and ref_count >= ref_sample_limit :\n",
    "                    diff_logodds = str(round(np.log(var_psi / (1.0 - var_psi)) - np.log(ref_psi / (1.0 - ref_psi)), 2))\n",
    "                    out_f.write(var_member['snpid'] + '\\t' + var_member['type'] + '\\t' + isoform + '\\t' + 'HOM' + '\\t' + str(aligned) + '\\t' + aligned_ref_seq + '\\t' + aligned_var_seq + '\\t' + str(ref_psi) + '\\t' + str(var_psi) + '\\t' + str(var_psi - ref_psi) + '\\t' + diff_logodds + '\\t' + str(var_member['count']) + '\\t' + str(mut_pos) + '\\t' + str(ref_count) + '\\t' + str(var_count) + '\\n')\n",
    "                elif np.abs(var_psi - ref_psi) >= 0.25 and var_count >= 2 and ref_count >= ref_sample_limit :\n",
    "                    diff_logodds = str(round(np.log(var_psi / (1.0 - var_psi)) - np.log(ref_psi / (1.0 - ref_psi)), 2))\n",
    "                    out_f.write(var_member['snpid'] + '\\t' + var_member['type'] + '\\t' + isoform + '\\t' + 'HOM' + '\\t' + str(aligned) + '\\t' + aligned_ref_seq + '\\t' + aligned_var_seq + '\\t' + str(ref_psi) + '\\t' + str(var_psi) + '\\t' + str(var_psi - ref_psi) + '\\t' + diff_logodds + '\\t' + str(var_member['count']) + '\\t' + str(mut_pos) + '\\t' + str(ref_count) + '\\t' + str(var_count) + '\\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         snp_id snp_type isoform  zyg  pas  \\\n",
      "0      chr1-:52818031-52818431:52818188/SNP/G/A      SNP       B  HOM    1   \n",
      "1  chr11+:10328480-10328880:10328747/INDEL/T/TC    INDEL       B  HOM    2   \n",
      "2      chr6+:71570645-71571045:71570880/SNP/T/A      SNP       B  HET    0   \n",
      "3      chr6+:71570645-71571045:71570880/SNP/T/A      SNP       B  HOM    0   \n",
      "4   chr7+:128589864-128590264:128590132/SNP/T/G      SNP       A  HET    0   \n",
      "\n",
      "                                             ref_seq  \\\n",
      "0  GGTCTACCCACCCAAGAGAAAAGACTGTTAACTGGAAGAAAAAATA...   \n",
      "1  AACTATCAATATTTAAGTTTGTTGCTGTCAAGATTTTTTTTGTAAC...   \n",
      "2  TGTGTTTACATTTTATGGTGCCTAGTATTGACAAAATGTTATTTCC...   \n",
      "3  TGTGTTTACATTTTATGGTGCCTAGTATTGACAAAATGTTATTTCC...   \n",
      "4  CCAGGTGCAGATGCCCAATCTTGATGCCCAGCCATCAGAACTGTGA...   \n",
      "\n",
      "                                             var_seq  ref_ratio  var_ratio  \\\n",
      "0  GGTCTACCCACCCAAGAGAAAAGACTGTTAACTGGAAGAAAAAATA...   0.794158   0.896942   \n",
      "1  AACTATCAATATTTAAGTTTGTTGCTGTCAAGATTTTTTTTGTAAC...   0.063214   0.246667   \n",
      "2  TGTGTTTACATTTTATGGTGCCTAGTATTGACAAAATGTTATTTCC...   0.649581   0.471429   \n",
      "3  TGTGTTTACATTTTATGGTGCCTAGTATTGACAAAATGTTATTTCC...   0.649581   0.215000   \n",
      "4  CCAGGTGCAGATGCCCAATCTTGATGCCCAGCCATCAGAACTGTGA...   0.312049   0.591667   \n",
      "\n",
      "       diff  diff_logodds  snp_count  \\\n",
      "0  0.102784          0.81          1   \n",
      "1  0.183452          1.58          1   \n",
      "2 -0.178153         -0.73          1   \n",
      "3 -0.434581         -1.91          1   \n",
      "4  0.279617          1.16          1   \n",
      "\n",
      "                                             snp_pos  ref_samples  var_samples  \n",
      "0                                                 81        101.0        121.0  \n",
      "1  121,122,123,127,129,130,131,132,133,134,135,13...         28.0          9.0  \n",
      "2                                                 79        382.0         28.0  \n",
      "3                                                 79        382.0          2.0  \n",
      "4                                                114        405.0         12.0  \n"
     ]
    }
   ],
   "source": [
    "\n",
    "df = pd.read_csv('APA_Tandem_UTR_GEUV_With_Id.csv', sep='\\t')\n",
    "\n",
    "print(df.head())\n",
    "\n",
    "df = df.sort_values(by='diff')\n",
    "\n",
    "df.to_csv('APA_Tandem_UTR_GEUV_With_Id_Sorted.csv', sep='\\t', header=True, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
