{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          PAS_ID Chromosome  Position Strand   Mean RPM Intron/exon location  \\\n",
      "0   chr1:16442:-       chr1   16442.0      -   2.433034               Intron   \n",
      "1  chr1:567556:-       chr1  567556.0      -   1.063078         3' most exon   \n",
      "2  chr1:755217:+       chr1  755217.0      +  10.742607         3' most exon   \n",
      "3  chr1:761586:-       chr1  761586.0      -   0.980149          Single exon   \n",
      "4  chr1:761586:-       chr1  761586.0      -   0.980149          Single exon   \n",
      "\n",
      "       Ensemble ID RefSeq Gene ID Gene Symbol  \\\n",
      "0  ENSG00000227232         653635      WASH7P   \n",
      "1               NO      102465432     MIR6723   \n",
      "2  ENSG00000177757         400728      FAM87B   \n",
      "3  ENSG00000225880          79854   LINC00115   \n",
      "4  ENSG00000225880          79854   LINC00115   \n",
      "\n",
      "                                     Gene Name Extension PAS type  \\\n",
      "0      WAS protein family homolog 7 pseudogene        NO       na   \n",
      "1                                microRNA 6723       YES       na   \n",
      "2  family with sequence similarity 87 member B       YES       na   \n",
      "3   long intergenic non-protein coding RNA 115        NO       na   \n",
      "4   long intergenic non-protein coding RNA 115       YES       na   \n",
      "\n",
      "              PSE PAS Signal Conservation  \n",
      "0           75.0%   OtherPAS           No  \n",
      "1  8.33333333333%      NoPAS           No  \n",
      "2  16.6666666667%     AAUAAA           No  \n",
      "3  29.1666666667%     AAUAAA           No  \n",
      "4  29.1666666667%     AAUAAA           No  \n"
     ]
    }
   ],
   "source": [
    "\n",
    "polyadb_df = pd.read_csv('polyadb/human.txt', delimiter='\\t')\n",
    "\n",
    "print(polyadb_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    chr   start     end         gene      feature strand       pas    mode\n",
      "0  chr1   16042   16842     WASH7P.1       Intron      -  OtherPAS   16442\n",
      "1  chr1  567156  567956    MIR6723.1  3_most_exon      -     NoPAS  567556\n",
      "2  chr1  754817  755617     FAM87B.1  3_most_exon      +    AAUAAA  755217\n",
      "3  chr1  761186  761986  LINC00115.1  Single_exon      -    AAUAAA  761586\n",
      "4  chr1  779551  780351  LINC01128.1       Intron      +     NoPAS  779951\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "chr_list = []\n",
    "start_list = []\n",
    "end_list = []\n",
    "gene_list = []\n",
    "strand_list = []\n",
    "feature_list = []\n",
    "pas_list = []\n",
    "mode_list = []\n",
    "\n",
    "gene_next_id = {}\n",
    "\n",
    "visited_dict = {}\n",
    "\n",
    "for index, row in polyadb_df.iterrows() :\n",
    "    \n",
    "    if row['PAS_ID'] in visited_dict :\n",
    "        continue\n",
    "    visited_dict[row['PAS_ID']] = True\n",
    "    \n",
    "    chr_list.append(row['Chromosome'])\n",
    "    start_list.append(int(row['Position']) - 400) #250\n",
    "    end_list.append(int(row['Position']) + 400)\n",
    "    \n",
    "    if row['Gene Symbol'] not in gene_next_id :\n",
    "        gene_next_id[row['Gene Symbol']] = 1\n",
    "    \n",
    "    gene_list.append(row['Gene Symbol'] + '.' + str(gene_next_id[row['Gene Symbol']]))\n",
    "    strand_list.append(row['Strand'])\n",
    "    feature_list.append(row['Intron/exon location'].replace(' ', '_').replace('\\'', ''))\n",
    "    pas_list.append(row['PAS Signal'])\n",
    "    mode_list.append(int(row['Position']))\n",
    "    \n",
    "    gene_next_id[row['Gene Symbol']] += 1\n",
    "\n",
    "\n",
    "polyadb_bed = pd.DataFrame({'chr'  : chr_list,\n",
    "                    'start'  : start_list,\n",
    "                    'end'  : end_list,\n",
    "                    'gene'  : gene_list,\n",
    "                    'strand'  : strand_list,\n",
    "                    'feature'  : feature_list,\n",
    "                    'pas' : pas_list,\n",
    "                    'mode' : mode_list,\n",
    "                })\n",
    "\n",
    "polyadb_bed = polyadb_bed[['chr', 'start', 'end', 'gene', 'feature', 'strand', 'pas', 'mode']]\n",
    "\n",
    "polyadb_bed.to_csv('polyadb_bed.bed', sep='\\t', header=False, index=False)\n",
    "\n",
    "print(polyadb_bed.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1  16042   16842   WASH7P.1     Intron       -  OtherPAS  16442\n",
      "chr1  567156  567956  MIR6723.1    3_most_exon  -  NoPAS     567556\n",
      "chr1  754817  755617  FAM87B.1     3_most_exon  +  AAUAAA    755217\n",
      "chr1  761186  761986  LINC00115.1  Single_exon  -  AAUAAA    761586\n",
      "chr1  779551  780351  LINC01128.1  Intron       +  NoPAS     779951\n",
      "\n",
      ">WASH7P.1\n",
      "TCTGGGATGGGAGCTGGGCCGGGGACCTCCCTGGTCACACACCTTCTTCCCTAGACACCCCACACTTTGTGTTTCAGACCTACAAGATGGGGTactaacaccacccccaccgcccccaccaccacccccaGCTCCTGAGGTGCTGGCCAGTGCACCCCCACTCCCACCCTCAACCGCGGCCCCTGTAGGCCAAGGCGCCAGGCAGGACGACAGCAGCAGCAGCGCGTCTCCTTCAGGTGGGAGCAGCTCTTTGAGGCCACCTGATTTCTGGCGTGCTCAGTGCACTCGGGTGGATTTTCTGTGGGTTTGTTAAGTGGTCAGAAATTCTCAATTTTTTGAATAGTTTCCATTTCAAATATCTTGTTCTACTTGGTTCATAAAATAGTGGTTTTCAAACTGTAGAGCTCTGGACTTCTCACTTCTAGGGCAGAGGGAGCCTGAACAAGTGAGGCTCTGGGTTCCCCATTCCTAATTAAACCAATGGAAAGAAGGGGTCTAATAACAAACTACAGCAACACATTTTTCATTTCAGCTTCACTGCTGTGTCTCCCAGTGTAACCCTAGCATCCAGAAGTGGCACAAAACCCCTCTGCTGGCTCGTGTGTGCAACTGAGACTGTCAGAGCATGGCTAGCTCAGGGGTCCAGCTCTGCAGGGTGGGGGCTAGAGAGGAAGCAGGGAGTATCTGCACACAGGATGCCCGCGCTCAGGTGGTTGCAGAAGTCAGTGCCCAGGCCCCCACACACAGTCTCCAAAGGTCCGGCCTCCCCAGCGCAGGGCTCCTCGTTTGAGGGGAGGTGA\n",
      ">MIR6723.1\n",
      "GGTGGGGGGCATCCATATAGTCACTCCAGGTTTATGGAGGGTTCTTCTACTATTAGGACTTTTCGCTTCGAAGCGAAGGCTTCTCAAATTATGAAAATTATTAATATTACTGCTGTTAGAGAAATGAATGAGCCTACAGATGATAGGATATTTCATGTGGTGTATGCATCGGGATAGTCCGAGTAACGTCGGGGCATTCCGGATAGGCCGAGAAAGTGTTGTGGGAAGAAAGTTAGATTTACGCCGATGAATATGATAGCGAAATGGATTTTGGCGTAGGTTTGGTCTAGGGTGTAGCCTGAGAATAGGGGAAATCAGTGAATGAAGCCTCCTATGATGGCAAATACAGCTCCTATTGATAGGACATAGTGGAAGTGGGCTACAACGTAGTACGTGTCGTGTAGTACGATGTCTAGTGATGAGTTTGCTAATACAATGCCAGTCAGGCCACCTACGGTGAAAAGAAAAATAAATCCTAGGGCTCAGAGCACTGCAGCAGATCATTTCATATTGCTTCCGTGGAGTGTGGCGAGTCAGCTAAATACTTTGACGCCGGTGGGGATAGCGATGATTATGGTAGCGGAGGTGAAATATGCTCGTGTGTCTACGTCTATTCCTACTGTAAATATATGGTGTGCTCACACGATAAACCCTAGGAAGCCAATTGATATCATAGCTCAGACCATACCTATGTATCCAAATGGTTCTTTTTTTCCCGGAGTAGTAAGTTACAATATGGGAGATTATTCCGAAGCCTGGTAGGATGAGAATATAAACTTCAGGGTGACCGAAAAATCAGA\n",
      ">FAM87B.1\n",
      "ACCGGCCAGTCTTATTCTCTAAAGTTGTTCCTTCTAAAGGGTGTGATGACCGAAAATTTCAGAAAAGCAAACCACCGCTGAAAGGCAACGTTATTTCTGTTGGCAGAAGGCGGCCTGAGCAATCTAGATTTTCCACGGTTCACCAACTAGTTTTTAAGGAAATATGGCTGTGagaggaataaaacatgattcctacctttaaggaactcagagAAGTGAATTAAAGGAAGTCACAGATCAGACAACCAACCACACAAAGTTTCTAAGAGCAAACTGTTCAGGTCGGCAAGTCActcttatccactgttttgccttctaaggtttcagttactctcagtcagtcatggtccaaaaacattaaatgaaaaattccagaaataaacaatacacacgtgttaaatcatgtttcattctgagtagcttgatgaagtctcatgccgtcccactcagccccacctggggtgtgacacctccctctgtcgagcagatccaccctgtctatactacctgcTTTTCCAGGAGATCCACCCTGTCTAGACTACCTGCGTGGCCAGCAGATCCACCCTATCTACACTACCTGCTTTTCCAGCAGATCCACCCTGTCTACACTACCTGCCTGTCCAGCAGATCAACACTGTCTACACTACCTGCTTTTCCAGCAGATCCACACTGTCTACACTACCTGCTTTTCCAGCAGATCCACCCCGTCTACACTACCTGCCTGGCCAGCATATCCACCCTGTCTACACTACCTGCTTTTCCAGTAGATCTGCCCTATCTACAATACCTGCTTGTCCAGC\n",
      ">LINC00115.1\n",
      "AGCAGAAGCCTAATTCCCAAATTCCTTAGTGGTTTTCTGATTTCCAAATTAGTTTCCCTTTTAAATTTATTGTGTCAGGTTCAGCTTATGAGGCCTCAATACTTTTCAGTCTTAATTGTATATTGAAAATACTTTTTGTTTACTAAATGCTTTTTACATTAATTCAGTGTGCACTCCGTAAGGATATTGATGATTTGAGTTAGTTTAGTATTCAACAGCTTCCTCTATTCCTTTGTATGATCTCTGTATTTAATGGCTGTGGCATAAAGTTTCCAACTAAGTATAAGTATCAAGTTTTCTTTGTGATGTTTTCTGCAAATATTGAAGGATGACCTGGATTGTCCTAGAACTTTGTTCCAACAGATTACATGTGTTCATAACGAATAAATTGCTCAAAGATATTTCCAAAGCTCACCTTTTATGTTTTTCAGTTCCAATAATTACATCTTTTTAAGGTTTATATTTTTTGATGACTTAATGTGATGTTCTGGAGAAGACAAATGCTTTTAATCAATATGATTAAAACCGTGAAAGACAAATCGCTGTTACTTAAGAGTGTGACATATGATCTGAAGTCTTTAGGGGGCAGGGATGTTAAGGGAAAACTGCCGCTATCGTATCAAGGTCATGCCATTCCTGTGAAGCTGGTGCTGGTTACTCTCTACTGGCTTCTGTTCACCTTTCCACCGGCCCCAAGACACACACATACTGACACAAGGCAGTGGATAGGGAAAGAGCAAGACTTTATAAAGTAAGCACAGAACAAAGGTATCTTTCTAATGAGGAGGAGGTGTTGGTTC\n",
      ">LINC01128.1\n",
      "cgtgtgaggacacagaggatggcatctgcacgccaaggagagaggcctcaggaggaaccagccctgcccacaccttgatctgggacttccagcctccagggttgtgagagagaaatgtctgttgtttaagccatgcagtctatggtgctttgttatggcagctccagcaagctaaCCTGGGATCTAAAATGATACTTAAACTGGTTCAGTCTGATTCTTAAACTGGTTCAGTGGAGCTAGGCCAGGCTCCCTCTGAGCAGATGTTGGTCCCTTTGCAGGCCTGCTGTGTGTGATGATTGCTAACCAGGTTATGACAGTGATCTTTAAGTAAGTGGGAATCTTGTGAGTTTTGAGGTTTTGTTTTTTTTTTTAATGAAATCTACATGGGATGTTTCAAAAATAGCAGCCTCTGTATACATTTGTCAAGCAGGAAGCTGGGTCTGCGGGGAGTAGGGTGGGGCTGGTTCTCCCACACGCTTGCCACTTTCAGGCCCTTGCCTTGCATGGGCTGCGGTGGTTCTGCCAGTGTGGATTCGAACCGATAGGTTTCGCTGTACCTTGATCAGCATCCTGCTGCCGAGAAAGCCAAAAAGGGCTGGAGCAACCCTCCCGTGTGCCATGTGAGCCTGTTTCCTTTCTACAGTCACCAGTTCACGCAGCAAGTGTCAGAGCATTCCTGTAAGGCCCGGCTCTCCACAGGGGCCAGGGGAGCTTTCTCTGGCAGGGTCTTCTTCACTGTGTGTAGGATACGGCAGCCTTtgaggctggggaaccgaagttttccttttattctattttaa\n",
      "\n"
     ]
    }
   ],
   "source": [
    "hg19_fai     = 'hg19/hg19.fa.fai'\n",
    "hg19_fa      = 'hg19/hg19.fa'\n",
    "\n",
    "# fasta\n",
    "output_fa = 'poladb_seqs.fa'\n",
    "output_bed = 'polyadb_bed.bed'\n",
    "!bedtools getfasta -name -s -fi \"$hg19_fa\" -bed \"$output_bed\" -fo \"$output_fa\"\n",
    "    \n",
    "# file tops\n",
    "!head -5 \"$output_bed\" | column -t ; echo\n",
    "!head -10 \"$output_fa\" ; echo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "85174\n",
      "{'seq': 'TCTGGGATGGGAGCTGGGCCGGGGACCTCCCTGGTCACACACCTTCTTCCCTAGACACCCCACACTTTGTGTTTCAGACCTACAAGATGGGGTACTAACACCACCCCCACCGCCCCCACCACCACCCCCAGCTCCTGAGGTGCTGGCCAGTGCACCCCCACTCCCACCCTCAACCGCGGCCCCTGTAGGCCAAGGCGCCAGGCAGGACGACAGCAGCAGCAGCGCGTCTCCTTCAGGTGGGAGCAGCTCTTTGAGGCCACCTGATTTCTGGCGTGCTCAGTGCACTCGGGTGGATTTTCTGTGGGTTTGTTAAGTGGTCAGAAATTCTCAATTTTTTGAATAGTTTCCATTTCAAATATCTTGTTCTACTTGGTTCATAAAATAGTGGTTTTCAAACTGTAGAGCTCTGGACTTCTCACTTCTAGGGCAGAGGGAGCCTGAACAAGTGAGGCTCTGGGTTCCCCATTCCTAATTAAACCAATGGAAAGAAGGGGTCTAATAACAAACTACAGCAACACATTTTTCATTTCAGCTTCACTGCTGTGTCTCCCAGTGTAACCCTAGCATCCAGAAGTGGCACAAAACCCCTCTGCTGGCTCGTGTGTGCAACTGAGACTGTCAGAGCATGGCTAGCTCAGGGGTCCAGCTCTGCAGGGTGGGGGCTAGAGAGGAAGCAGGGAGTATCTGCACACAGGATGCCCGCGCTCAGGTGGTTGCAGAAGTCAGTGCCCAGGCCCCCACACACAGTCTCCAAAGGTCCGGCCTCCCCAGCGCAGGGCTCCTCGTTTGAGGGGAGGTGA', 'chr': 'chr1', 'seq_start': 16042, 'seq_end': 16842, 'strand': '-', 'type': 'Intron', 'pas': 'OtherPAS', 'cut_mode': 16442, 'sort_order': 0}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "seq_dict = {}\n",
    "with open('poladb_seqs.fa', 'r') as f :\n",
    "    i = 0\n",
    "    seq_id = ''\n",
    "    for line in f :\n",
    "        if i % 2 == 0 :\n",
    "            seq_id = line[1:-1]\n",
    "        else :\n",
    "            seq_dict[seq_id] = line[:-1].upper()\n",
    "        \n",
    "        i += 1\n",
    "\n",
    "gene_dict = {}\n",
    "\n",
    "df_dict = {}\n",
    "for index, row in polyadb_bed.iterrows() :\n",
    "    df_dict[row['gene']] = {}\n",
    "    \n",
    "    df_dict[row['gene']]['seq'] = seq_dict[row['gene']]\n",
    "    df_dict[row['gene']]['chr'] = row['chr']\n",
    "    df_dict[row['gene']]['seq_start'] = row['start']\n",
    "    df_dict[row['gene']]['seq_end'] = row['end']\n",
    "    df_dict[row['gene']]['strand'] = row['strand']\n",
    "    df_dict[row['gene']]['type'] = row['feature']\n",
    "    df_dict[row['gene']]['pas'] = row['pas']\n",
    "    df_dict[row['gene']]['cut_mode'] = row['mode']\n",
    "    df_dict[row['gene']]['sort_order'] = index\n",
    "    \n",
    "    gene_id = row['gene'].split('.')[0]\n",
    "    \n",
    "    if gene_id not in gene_dict :\n",
    "        gene_dict[gene_id] = 0\n",
    "    gene_dict[gene_id] += 1\n",
    "\n",
    "\n",
    "print(len(df_dict))\n",
    "print(df_dict['WASH7P.1'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Make Valid PAS lookup hierarchy\n",
    "\n",
    "cano_pas1 = 'AATAAA'\n",
    "cano_pas2 = 'ATTAAA'\n",
    "\n",
    "valid_pas = []\n",
    "\n",
    "valid_pas.append({})\n",
    "valid_pas[0]['AATAAA'] = True\n",
    "\n",
    "valid_pas.append({})\n",
    "valid_pas[1]['ATTAAA'] = True\n",
    "\n",
    "valid_pas.append({})\n",
    "valid_pas[2]['AGTAAA'] = True\n",
    "valid_pas[2]['TATAAA'] = True\n",
    "valid_pas[2]['CATAAA'] = True\n",
    "valid_pas[2]['GATAAA'] = True\n",
    "\n",
    "valid_pas.append({})\n",
    "for pos in range(0, 6) :\n",
    "    for base in ['A', 'C', 'G', 'T'] :\n",
    "        valid_pas[3][cano_pas1[:pos] + base + cano_pas1[pos+1:]] = True\n",
    "\n",
    "valid_pas.append({})\n",
    "for pos1 in range(0, 6) :\n",
    "    for pos2 in range(pos1 + 1, 6) :\n",
    "        for base1 in ['A', 'C', 'G', 'T'] :\n",
    "            for base2 in ['A', 'C', 'G', 'T'] :\n",
    "                valid_pas[4][cano_pas1[:pos1] + base1 + cano_pas1[pos1+1:pos2] + base2 + cano_pas1[pos2+1:]] = True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def search_seq(seq, cut_start, cut_end, search_pas, before_cut = 45, after_cut = 1) :\n",
    "    \n",
    "    align_j = cut_start - 25\n",
    "    aligned = -1\n",
    "    \n",
    "    for j in range(cut_start - before_cut, cut_start + after_cut) :\n",
    "\n",
    "        candidate_pas = seq[j:j+6]\n",
    "\n",
    "        if candidate_pas == search_pas :\n",
    "            align_j = j\n",
    "            aligned = 0\n",
    "        \n",
    "        if aligned != -1 :\n",
    "            break\n",
    "    \n",
    "    aligned_seq = (seq[align_j-50:])[:186]\n",
    "    aligned_seq_long = (seq[align_j-100:])[:186 + 50 + 50]\n",
    "    \n",
    "    if len(aligned_seq_long) != 186 + 50 + 50 :\n",
    "        print('WARNING! Aligned seq not long enough. Length = ' + str(len(aligned_seq_long)))\n",
    "    \n",
    "    \n",
    "    aligned_wide_seq = (seq[align_j-125:])[:256]\n",
    "    aligned_wide_seq_long = (seq[align_j-175:])[:256 + 50 + 50]\n",
    "    \n",
    "    if len(aligned_wide_seq_long) != 256 + 50 + 50 :\n",
    "        print('WARNING! Aligned seq not long enough. Length = ' + str(len(aligned_wide_seq_long)))\n",
    "    \n",
    "    return aligned_seq, aligned_seq_long, aligned, (align_j - cut_start), aligned_wide_seq, aligned_wide_seq_long\n",
    "\n",
    "def align_seq(seq, cut_start, cut_end, before_cut = 45, after_cut = 1) :\n",
    "    \n",
    "    align_j = cut_start - 25\n",
    "    aligned = -1\n",
    "    \n",
    "    for i in range(0, len(valid_pas)) :\n",
    "        for j in range(cut_start - before_cut, cut_start + after_cut) :\n",
    "\n",
    "            candidate_pas = seq[j:j+6]\n",
    "\n",
    "            if candidate_pas in valid_pas[i] :\n",
    "                align_j = j\n",
    "                aligned = i\n",
    "                \n",
    "                if j >= cut_start :\n",
    "                    break\n",
    "        if aligned != -1 :\n",
    "            break\n",
    "    \n",
    "    aligned_seq = (seq[align_j-50:])[:186]\n",
    "    aligned_seq_long = (seq[align_j-100:])[:186 + 50 + 50]\n",
    "    \n",
    "    if len(aligned_seq_long) != 186 + 50 + 50 :\n",
    "        print('WARNING! Aligned seq not long enough. Length = ' + str(len(aligned_seq_long)))\n",
    "    \n",
    "    \n",
    "    aligned_wide_seq = (seq[align_j-125:])[:256]\n",
    "    aligned_wide_seq_long = (seq[align_j-175:])[:256 + 50 + 50]\n",
    "    \n",
    "    if len(aligned_wide_seq_long) != 256 + 50 + 50 :\n",
    "        print('WARNING! Aligned seq not long enough. Length = ' + str(len(aligned_wide_seq_long)))\n",
    "    \n",
    "    return aligned_seq, aligned_seq_long, aligned, (align_j - cut_start), aligned_wide_seq, aligned_wide_seq_long\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         gene    gene_id  sitenum  num_sites  pas  \\\n",
      "673   AADACL3  AADACL3.1        1          2    0   \n",
      "674   AADACL3  AADACL3.2        2          2    0   \n",
      "3644    ABCA4    ABCA4.2        1          2    2   \n",
      "3643    ABCA4    ABCA4.1        2          2    0   \n",
      "7417   ABCB10   ABCB10.2        1          2    0   \n",
      "\n",
      "                                                    seq  \\\n",
      "673   GTGCTTTCTAATCTGTGGAATGCCAGGGTCCCAGTGTGGGAGCCTT...   \n",
      "674   GCTGCAGGTGGTGGTTGCTGAAGGTGGGGGAGGCTGTGGCAATTTC...   \n",
      "3644  GTTTATCAAATACAACTCAGACGTCAGTCTCCTGGCCCCTTTGAGA...   \n",
      "3643  TTTCTGCATGTTTGTCTGTGTGTCTGCGTTGTGTGTGATTTTCATG...   \n",
      "7417  TCTGATACATGATGTTCAATTTTATCTTTAGGTAATATTTTATATC...   \n",
      "\n",
      "                                                seq_ext  \\\n",
      "673   TTACCCCCAGGGAATTAGGGGAGAGGAAACACCTTTATTTGCTTTC...   \n",
      "674   TAAATTGTCGTAATCTTTTTGCTGGTTGATGGTCTTGCCTTGATGT...   \n",
      "3644  CCTCTGATCCATCTCTGTCTGCATGAGTGACAGCTGGCAGAGTCCT...   \n",
      "3643  CCACTAGCAGCTTTGGCCTCCATATTGCTCTCATTTCAAGCAGATC...   \n",
      "7417  TGTGCCATAGAAGTATTTACGAAATTGCATTTCATTGTTATGTTTT...   \n",
      "\n",
      "                                               wide_seq  \\\n",
      "673   TGAGACAATTTTACTGTCTTAGTTATTACCCCCAGGGAATTAGGGG...   \n",
      "674   TGTATGTTACATTCATGGGAATGTCTAAATTGTCGTAATCTTTTTG...   \n",
      "3644  ATCTCCTCCTGTCATCTCCACCAGGCCTCTGATCCATCTCTGTCTG...   \n",
      "3643  CCAGAACTAGAAACCCCGGGCCATCCCACTAGCAGCTTTGGCCTCC...   \n",
      "7417  GGTTGAATCTGAGGAAAATAATCCTTGTGCCATAGAAGTATTTACG...   \n",
      "\n",
      "                                           wide_seq_ext    pas_pos   cut_mode  \\\n",
      "673   AGACAGACAGGAAGATTTGAGAAAAATCAATGAGAGGAAAAAGTCA...   12780316   12780335   \n",
      "674   TGCGTGGGTTTTCTCTGAGTTCTCCAGCTTCCTCCCACATTCCAAA...   12788705   12788725   \n",
      "3644  GCTCCTCATGCTGTATCCCCAGTCTCTCGGCCTGCCATGTCATCAT...   94501567   94501545   \n",
      "3643  TGGAACACCTGATGGTGAAACCAAACAAATACAAAATCCTTCTCCA...   94458420   94458384   \n",
      "7417  TGTGTTCTTTATAAAGTGTGATTTTCAGAAAGCAAACAACACAATT...  229652351  229652329   \n",
      "\n",
      "     chrom strand    site_type  \n",
      "673   chr1      +       Intron  \n",
      "674   chr1      +  3_most_exon  \n",
      "3644  chr1      -       Intron  \n",
      "3643  chr1      -  3_most_exon  \n",
      "7417  chr1      -  3_most_exon  \n",
      "         gene   gene_id  sitenum  num_sites  pas  \\\n",
      "81499  TXLNGY  TXLNGY.3        3          5    0   \n",
      "81500  TXLNGY  TXLNGY.4        4          5    3   \n",
      "81501  TXLNGY  TXLNGY.5        5          5    0   \n",
      "81481   USP9Y   USP9Y.1        1          1    0   \n",
      "81509   VAMP7   VAMP7.2        2          2   -1   \n",
      "\n",
      "                                                     seq  \\\n",
      "81499  TGTTTCTGGACACTGTAGCTGACCAAGAAAATGTTTAAATGTTGCG...   \n",
      "81500  TTATCTACTCAACTTCGTTCAGGTTAACTAAAGATATAGATATAAT...   \n",
      "81501  GATGGGCCCTTGGGTTGGTTCCAAGTATTTGCTATTGTAAACAGTG...   \n",
      "81481  TTAAATGTTTGGCTGTTTTGTTTTATAGAAATTATTTTGCTGAATT...   \n",
      "81509  TGGTGCTGATAGTCTCATCATGGGGTTTCCACAGACCTTCGATTTG...   \n",
      "\n",
      "                                                 seq_ext  \\\n",
      "81499  GAGTTGAACATGTCCATAAATGTATACAATCTGACCCTGTTTTTTG...   \n",
      "81500  TTTTTGTTTTTGTTTTTGTTTTTGTTTTTTTACCTAATGTGAATCT...   \n",
      "81501  CCTATGTCCTGAATGGTATGTGCCAAATTTTCTTTATCCAGTCTAT...   \n",
      "81481  GTTTAAAAACCAAATATACCTGCAAATAGATACAGCCTATCCTATA...   \n",
      "81509  AAACAAAAAGTGCAGGCCAAGCAGGAAAAAAAAAAAACAAACCAGA...   \n",
      "\n",
      "                                                wide_seq  \\\n",
      "81499  AACCTCTCTGTTTGCCAAGCCTGCAGAGTTGAACATGTCCATAAAT...   \n",
      "81500  ACTTCTGTAGAGTTTGTGTTTTTTGTTTTTGTTTTTGTTTTTGTTT...   \n",
      "81501  TTCAGACATGAGGTCTTTGCCCGTGCCTATGTCCTGAATGGTATGT...   \n",
      "81481  GAATGTGTGTAATATTTTGGAACCTGTTTAAAAACCAAATATACCT...   \n",
      "81509  AAAGTATAATAATAATAATAAAAAGAAACAAAAAGTGCAGGCCAAG...   \n",
      "\n",
      "                                            wide_seq_ext   pas_pos  cut_mode  \\\n",
      "81499  GGCCCTGTGGCATGCCCAGTATCTGCAATGTTCTAGAATAGAAGTT...  21767679  21767701   \n",
      "81500  TTTTTTAAGGATATGATTTTATAAGAGGCAGTTCTCTTTTGCTTTT...  21768158  21768160   \n",
      "81501  GTTAGATTCCATTTGACAATTTTGGCTTTTGTCGCCATTGCTTTTG...  21771708  21771731   \n",
      "81481  CAAGAAACCTGTGAGTTAATACTATATTTAAAAGAGAATTGGTAAA...  14972747  14972765   \n",
      "81509  GTATACCTGTGTAAGAAACCTGCACGTTCTGCACATGTATCCTAGA...  59260153  59260178   \n",
      "\n",
      "      chrom strand    site_type  \n",
      "81499  chrY      +  3_most_exon  \n",
      "81500  chrY      +  3_most_exon  \n",
      "81501  chrY      +  3_most_exon  \n",
      "81481  chrY      +  3_most_exon  \n",
      "81509  chrY      +       Intron  \n",
      "Total number of members: 85174\n"
     ]
    }
   ],
   "source": [
    "#Global PolyaDB dataframe generation\n",
    "\n",
    "gene = []\n",
    "\n",
    "prox_id = []\n",
    "prox_seq = []\n",
    "prox_seq_ext = []\n",
    "prox_wide_seq = []\n",
    "prox_wide_seq_ext = []\n",
    "\n",
    "cut_mode = []\n",
    "\n",
    "prox_type = []\n",
    "prox_pas = []\n",
    "prox_sitenum = []\n",
    "num_sites = []\n",
    "\n",
    "prox_chrom = []\n",
    "prox_strand = []\n",
    "\n",
    "pas_pos = []\n",
    "\n",
    "for site_id in df_dict :\n",
    "    \n",
    "    gene_id = site_id.split('.')[0]\n",
    "    site_num = int(site_id.split('.')[1])\n",
    "    \n",
    "    gene.append(gene_id)\n",
    "    \n",
    "    p_id = site_id\n",
    "    \n",
    "    p_seq = df_dict[p_id]['seq']\n",
    "    \n",
    "    p_mode = df_dict[p_id]['cut_mode']\n",
    "    \n",
    "    p_type = df_dict[p_id]['type']\n",
    "    \n",
    "    p_pas = df_dict[p_id]['pas'].replace('U', 'T')\n",
    "    \n",
    "    if p_pas in ['NoPAS', 'OtherPAS'] :\n",
    "        p_seq, p_seq_long, p_aligned, p_shift, p_wide_seq, p_wide_seq_long = align_seq(p_seq, 400, len(p_seq) - 800, before_cut = 45, after_cut = 1) #250, ..., 500\n",
    "    else :\n",
    "        p_seq, p_seq_long, p_aligned, p_shift, p_wide_seq, p_wide_seq_long = search_seq(p_seq, 400, len(p_seq) - 800, p_pas, before_cut = 45, after_cut = 1)\n",
    "    \n",
    "    if df_dict[p_id]['strand'] == '+' :\n",
    "        pas_pos.append(p_shift + p_mode)\n",
    "    else :\n",
    "        pas_pos.append(-p_shift + p_mode)\n",
    "    \n",
    "    prox_id.append(p_id)\n",
    "    prox_seq.append(p_seq)\n",
    "    prox_seq_ext.append(p_seq_long)\n",
    "    \n",
    "    prox_wide_seq.append(p_wide_seq)\n",
    "    prox_wide_seq_ext.append(p_wide_seq_long)\n",
    "    \n",
    "    cut_mode.append(p_mode)\n",
    "    \n",
    "    prox_type.append(p_type)\n",
    "    \n",
    "    prox_pas.append(p_aligned)\n",
    "    \n",
    "    prox_chrom.append(df_dict[p_id]['chr'])\n",
    "    prox_strand.append(df_dict[p_id]['strand'])\n",
    "    \n",
    "    #sort_order.append(df_dict[p_id]['sort_order'])\n",
    "    p_sitenum = -1\n",
    "    if df_dict[p_id]['strand'] == '-' :\n",
    "        p_sitenum = gene_dict[gene_id] - site_num + 1\n",
    "    elif df_dict[p_id]['strand'] == '+' :\n",
    "        p_sitenum = site_num\n",
    "    \n",
    "    num_sites.append(gene_dict[gene_id])\n",
    "    prox_sitenum.append(p_sitenum)\n",
    "\n",
    "\n",
    "df = pd.DataFrame({\n",
    "        'gene' : gene,\n",
    "        'gene_id'  : prox_id,\n",
    "        'sitenum'  : prox_sitenum,\n",
    "        'num_sites' : num_sites,\n",
    "        'pas'  : prox_pas,\n",
    "        'seq'  : prox_seq,\n",
    "        'seq_ext'  : prox_seq_ext,\n",
    "        'wide_seq'  : prox_wide_seq,\n",
    "        'wide_seq_ext'  : prox_wide_seq_ext,\n",
    "        'pas_pos'  : pas_pos,\n",
    "        'cut_mode'  : cut_mode,\n",
    "        'chrom' : prox_chrom,\n",
    "        'strand' : prox_strand,\n",
    "        'site_type'  : prox_type\n",
    "    })\n",
    "\n",
    "df = df[['gene',\n",
    "        'gene_id',\n",
    "        'sitenum',\n",
    "        'num_sites',\n",
    "        'pas',\n",
    "        'seq',\n",
    "        'seq_ext',\n",
    "        'wide_seq',\n",
    "        'wide_seq_ext',\n",
    "        'pas_pos',\n",
    "        'cut_mode',\n",
    "        'chrom',\n",
    "        'strand',\n",
    "        'site_type']]\n",
    "\n",
    "df = df.sort_values(by=['chrom', 'gene', 'sitenum'])\n",
    "\n",
    "print(df.head())\n",
    "print(df.tail())\n",
    "\n",
    "print('Total number of members: ' + str(len(df)))\n",
    "\n",
    "\n",
    "df.to_csv('polyadb_processed_v3.csv', header=True, index=False, sep=',')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         gene chrom strand proximal_id  distal_id  proximal_sitenum  \\\n",
      "507   AADACL3  chr1      +   AADACL3.1  AADACL3.2                 1   \n",
      "2815    ABCA4  chr1      -     ABCA4.2    ABCA4.1                 1   \n",
      "5752   ABCB10  chr1      -    ABCB10.2   ABCB10.1                 1   \n",
      "2825    ABCD3  chr1      +     ABCD3.1    ABCD3.2                 1   \n",
      "2826    ABCD3  chr1      +     ABCD3.2    ABCD3.3                 2   \n",
      "\n",
      "      distal_sitenum  proximal_pas  distal_pas  \\\n",
      "507                2             0           0   \n",
      "2815               2             2           0   \n",
      "5752               2             0           0   \n",
      "2825               2            -1           0   \n",
      "2826               3             0           2   \n",
      "\n",
      "                                           proximal_seq  \\\n",
      "507   GTGCTTTCTAATCTGTGGAATGCCAGGGTCCCAGTGTGGGAGCCTT...   \n",
      "2815  GTTTATCAAATACAACTCAGACGTCAGTCTCCTGGCCCCTTTGAGA...   \n",
      "5752  TCTGATACATGATGTTCAATTTTATCTTTAGGTAATATTTTATATC...   \n",
      "2825  TTCTTGAATTTTATAAGTATCTCTAGCTTCTTGATAACTTATAAGG...   \n",
      "2826  TTCGAGACAAGCCTGGACAAAAAGCGAGACCCGCTTCTTTAAAAAA...   \n",
      "\n",
      "                                             distal_seq  site_distance  \\\n",
      "507   GCTGCAGGTGGTGGTTGCTGAAGGTGGGGGAGGCTGTGGCAATTTC...           8390   \n",
      "2815  TTTCTGCATGTTTGTCTGTGTGTCTGCGTTGTGTGTGATTTTCATG...          43161   \n",
      "5752  TCAGGAATAAAGAAAAGACTAACATTACACATATCCAAAAACATGT...            190   \n",
      "2825  TTCGAGACAAGCCTGGACAAAAAGCGAGACCCGCTTCTTTAAAAAA...           1468   \n",
      "2826  GGAGTGCATTTGACTCCAGGAAAAGCCATTTTGGTTTTCCTTAACT...          39040   \n",
      "\n",
      "      proximal_cut_mode  distal_cut_mode proximal_type  distal_type  num_sites  \n",
      "507            12780335         12788725        Intron  3_most_exon          2  \n",
      "2815           94501545         94458384        Intron  3_most_exon          2  \n",
      "5752          229652329        229652139   3_most_exon  3_most_exon          2  \n",
      "2825           94942811         94944279        Intron       Intron          8  \n",
      "2826           94944279         94983319        Intron  3_most_exon          8  \n",
      "         gene chrom strand proximal_id distal_id  proximal_sitenum  \\\n",
      "63607  TTTY14  chrY      -    TTTY14.2  TTTY14.1                 3   \n",
      "63611  TXLNGY  chrY      +    TXLNGY.1  TXLNGY.2                 1   \n",
      "63612  TXLNGY  chrY      +    TXLNGY.2  TXLNGY.3                 2   \n",
      "63613  TXLNGY  chrY      +    TXLNGY.3  TXLNGY.4                 3   \n",
      "63614  TXLNGY  chrY      +    TXLNGY.4  TXLNGY.5                 4   \n",
      "\n",
      "       distal_sitenum  proximal_pas  distal_pas  \\\n",
      "63607               4             4           0   \n",
      "63611               2             0           2   \n",
      "63612               3             2           0   \n",
      "63613               4             0           3   \n",
      "63614               5             3           0   \n",
      "\n",
      "                                            proximal_seq  \\\n",
      "63607  CATCTATTTTGGATTGTATATTGTGTTTGTGTATTTACGCTTTGAT...   \n",
      "63611  CCATGTTGTAAATGCACAAAACAAATATTTAGTTAATAATCAACTC...   \n",
      "63612  TCAAATACTAATTTAAACTCTGGAGCCCATAATATCTACATAAAAG...   \n",
      "63613  TGTTTCTGGACACTGTAGCTGACCAAGAAAATGTTTAAATGTTGCG...   \n",
      "63614  TTATCTACTCAACTTCGTTCAGGTTAACTAAAGATATAGATATAAT...   \n",
      "\n",
      "                                              distal_seq  site_distance  \\\n",
      "63607  GATTCATAGTAACTTCTTATGGAATTGATTTGCATTGAACACAAAC...             69   \n",
      "63611  TCAAATACTAATTTAAACTCTGGAGCCCATAATATCTACATAAAAG...          14075   \n",
      "63612  TGTTTCTGGACACTGTAGCTGACCAAGAAAATGTTTAAATGTTGCG...           1321   \n",
      "63613  TTATCTACTCAACTTCGTTCAGGTTAACTAAAGATATAGATATAAT...            459   \n",
      "63614  GATGGGCCCTTGGGTTGGTTCCAAGTATTTGCTATTGTAAACAGTG...           3571   \n",
      "\n",
      "       proximal_cut_mode  distal_cut_mode  proximal_type  distal_type  \\\n",
      "63607           21152582         21152513         Intron       Intron   \n",
      "63611           21752305         21766380  Internal_exon  3_most_exon   \n",
      "63612           21766380         21767701    3_most_exon  3_most_exon   \n",
      "63613           21767701         21768160    3_most_exon  3_most_exon   \n",
      "63614           21768160         21771731    3_most_exon  3_most_exon   \n",
      "\n",
      "       num_sites  \n",
      "63607          4  \n",
      "63611          5  \n",
      "63612          5  \n",
      "63613          5  \n",
      "63614          5  \n",
      "Total number of members: 63817\n"
     ]
    }
   ],
   "source": [
    "#Pairwise PolyaDB dataframe generation\n",
    "\n",
    "chrom = []\n",
    "strand = []\n",
    "gene = []\n",
    "\n",
    "prox_id = []\n",
    "dist_id = []\n",
    "\n",
    "prox_seq = []\n",
    "dist_seq = []\n",
    "\n",
    "site_distance = []\n",
    "\n",
    "prox_cut_mode = []\n",
    "dist_cut_mode = []\n",
    "\n",
    "prox_type = []\n",
    "dist_type = []\n",
    "\n",
    "prox_pas = []\n",
    "dist_pas = []\n",
    "\n",
    "prox_sitenum = []\n",
    "dist_sitenum = []\n",
    "\n",
    "num_sites = []\n",
    "\n",
    "for site_id in df_dict :\n",
    "    \n",
    "    gene_id = site_id.split('.')[0]\n",
    "    site_num = int(site_id.split('.')[1])\n",
    "    \n",
    "    p_id = site_id\n",
    "    d_id = ''\n",
    "    if df_dict[site_id]['strand'] == '-' :\n",
    "        if gene_id + '.' + str(site_num - 1) not in df_dict :\n",
    "            continue\n",
    "        \n",
    "        d_id = gene_id + '.' + str(site_num - 1)\n",
    "    elif df_dict[site_id]['strand'] == '+' :\n",
    "        if gene_id + '.' + str(site_num + 1) not in df_dict :\n",
    "            continue\n",
    "        \n",
    "        d_id = gene_id + '.' + str(site_num + 1)\n",
    "    \n",
    "    \n",
    "    gene.append(gene_id)\n",
    "    chrom.append(df_dict[p_id]['chr'])\n",
    "    strand.append(df_dict[p_id]['strand'])\n",
    "    \n",
    "    p_seq = df_dict[p_id]['seq']\n",
    "    p_mode = df_dict[p_id]['cut_mode']\n",
    "    p_type = df_dict[p_id]['type']\n",
    "    p_pas = df_dict[p_id]['pas'].replace('U', 'T')\n",
    "    \n",
    "    if p_pas in ['NoPAS', 'OtherPAS'] :\n",
    "        p_seq, p_seq_long, p_aligned, p_shift, _, _ = align_seq(p_seq, 400, len(p_seq) - 800, before_cut = 45, after_cut = 1) #250, ..., 500\n",
    "    else :\n",
    "        p_seq, p_seq_long, p_aligned, p_shift, _, _ = search_seq(p_seq, 400, len(p_seq) - 800, p_pas, before_cut = 45, after_cut = 1)\n",
    "    \n",
    "    prox_id.append(p_id)\n",
    "    prox_seq.append(p_seq)\n",
    "    prox_cut_mode.append(p_mode)\n",
    "    prox_type.append(p_type)\n",
    "    prox_pas.append(p_aligned)\n",
    "    \n",
    "    d_seq = df_dict[d_id]['seq']\n",
    "    d_mode = df_dict[d_id]['cut_mode']\n",
    "    d_type = df_dict[d_id]['type']\n",
    "    d_pas = df_dict[d_id]['pas'].replace('U', 'T')\n",
    "    \n",
    "    if d_pas in ['NoPAS', 'OtherPAS'] :\n",
    "        d_seq, d_seq_long, d_aligned, d_shift = align_seq(d_seq, 250, len(d_seq) - 500, before_cut = 45, after_cut = 1)\n",
    "    else :\n",
    "        d_seq, d_seq_long, d_aligned, d_shift = search_seq(d_seq, 250, len(d_seq) - 500, d_pas, before_cut = 45, after_cut = 1)\n",
    "    \n",
    "    dist_id.append(d_id)\n",
    "    dist_seq.append(d_seq)\n",
    "    dist_cut_mode.append(d_mode)\n",
    "    dist_type.append(d_type)\n",
    "    dist_pas.append(d_aligned)\n",
    "    \n",
    "    if df_dict[site_id]['strand'] == '-' :\n",
    "        prox_sitenum.append(gene_dict[gene_id] - site_num + 1)\n",
    "        dist_sitenum.append(gene_dict[gene_id] - site_num + 2)\n",
    "    elif df_dict[site_id]['strand'] == '+' :\n",
    "        prox_sitenum.append(site_num)\n",
    "        dist_sitenum.append(site_num + 1)\n",
    "    \n",
    "    site_dist = numpy.abs(df_dict[p_id]['cut_mode'] - df_dict[d_id]['cut_mode'])\n",
    "    site_distance.append(site_dist)\n",
    "    num_sites.append(gene_dict[gene_id])\n",
    "        \n",
    "\n",
    "\n",
    "df = pd.DataFrame({\n",
    "        'gene' : gene,\n",
    "        'chrom' : chrom,\n",
    "        'strand' : strand,\n",
    "        'proximal_id'  : prox_id,\n",
    "        'distal_id'  : dist_id,\n",
    "        'proximal_sitenum'  : prox_sitenum,\n",
    "        'distal_sitenum'  : dist_sitenum,\n",
    "        'proximal_pas'  : prox_pas,\n",
    "        'distal_pas'  : dist_pas,\n",
    "        'proximal_seq'  : prox_seq,\n",
    "        'distal_seq'  : dist_seq,\n",
    "        'site_distance'  : site_distance,\n",
    "        'proximal_cut_mode'  : prox_cut_mode,\n",
    "        'distal_cut_mode'  : dist_cut_mode,\n",
    "        'proximal_type'  : prox_type,\n",
    "        'distal_type'  : dist_type,\n",
    "        'num_sites' : num_sites\n",
    "    })\n",
    "\n",
    "df = df[['gene',\n",
    "        'chrom',\n",
    "        'strand',\n",
    "        'proximal_id',\n",
    "        'distal_id',\n",
    "        'proximal_sitenum',\n",
    "        'distal_sitenum',\n",
    "        'proximal_pas',\n",
    "        'distal_pas',\n",
    "        'proximal_seq',\n",
    "        'distal_seq',\n",
    "        'site_distance',\n",
    "        'proximal_cut_mode',\n",
    "        'distal_cut_mode',\n",
    "        'proximal_type',\n",
    "        'distal_type',\n",
    "        'num_sites']]\n",
    "\n",
    "df = df.sort_values(by=['chrom', 'gene', 'proximal_sitenum'])\n",
    "\n",
    "print(df.head())\n",
    "print(df.tail())\n",
    "\n",
    "print('Total number of members: ' + str(len(df)))\n",
    "\n",
    "\n",
    "df.to_csv('polyadb_processed_v3_pairs.csv', header=True, index=False, sep=',')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:tensorflow]",
   "language": "python",
   "name": "conda-env-tensorflow-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
